### ISSUE TITLE
Question: Tpetra: Appropriate data structure for block data

### ISSUE THREAD

**Issue by ctanis** [trilinos/Trilinos#2518]
I am getting terrible performance with Experimental::BlockCrsMatrix  and MultiVec (BlockCrsVector doesn't compile).   What is currently the best way to deal with multiple degrees of freedom per mesh point?

@trilinos/tpetra

## Expectations

Experimental::BlockCrsMatrix.apply should be faster in parallel, or I should be doing this in a different way.

## Current Behavior

Using MPI (np = 2), BlockCrsMatrix.apply is taking approximately 100 times longer than in serial (still getting the correct answer, I think).

This is applying a block matrix with 1 degree of freedom to a MultiVector (not block) with 1 vector.

Attempting to compile a simple example with BlockVector does not compile. 
[compilation_error.txt](https://github.com/trilinos/Trilinos/files/1891118/compilation_error.txt)

## Motivation and Context

I understand these are "experimental" data structures, so I would love to know the most appropriate way to go forward.

Also note:  I was using 1 degree of freedom per mesh point to make sure I was solving the problem correctly.  My imminent need is to have 4 or more per mesh point.  Potentially many more!

I'm going to attempt to add a 'question' label after submitting this.  My github interface is somehow missing the label and assignee dropdowns.

**Comment by mhoemmen** [comment]
@ctanis What version of Trilinos are you using?  Also, are you using OpenMP?

**Comment by mhoemmen** [comment]
@trilinos/tpetra

**Comment by mhoemmen** [comment]
> My github interface is somehow missing the label and assignee dropdowns.

The Trilinos repository restricts assignment of labels and assignees to members.

**Comment by kyungjoo-kim** [comment]
@ctanis We have an example code to measure performance of block crs matrix vector multiplication. 
After you compile Trilinos, you can find out an executable ``packages/tpetra/core/example/BlockCrs/TpetraCore_BlockCrsPerfTest.exe``. 
Testing the code with 
- 7 point 3D stencil on a 32x32x32 block mesh 
- my workstation xeon 2x8 intel sandy bridge
- OMP_NUM_THREADS=1 (mpi only).
I have the following result. Could you verify that you also get the same performance as I have seen below ? The code compares the block crs apply method against the point crs apply method. 

```
[kyukim @bread] BlockCrs > mpirun -np 1 ./TpetraCore_BlockCrsPerfTest.exe --num-elements-i=32 --num-elements-j=32 --num-elements-k=32 --num-procs-i=1 --num-procs-j=1 --num-procs-k=1
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Kokkos::OpenMP thread_pool_topology[ 1 x 1 x 1 ]
Column = 0  Error norm = 0
====================================================================================

                        TimeMonitor results over 1 processor

Timer Name                                               Global time (num calls)
------------------------------------------------------------------------------------
0) LocalGraphConstruction                                0.04422 (1)
1) GlobalGraphConstruction                               0.002842 (1)
2) LocalBlockCrsFill                                     0.3604 (1)
3) BlockCrsMatrix FillComplete - currently do nothing    9.537e-07 (1)
4) MultiVectorFill                                       0.006302 (1)
5) BlockCrs Apply                                        7.918 (100)
6) Conversion from BlockCrs to PointCrs                  0.155 (1)
7) PointCrs Apply                                        10.02 (100)
8) Export MatrixMarket                                   21.14 (1)
X) Global                                                39.71 (1)
====================================================================================
[kyukim @bread] BlockCrs > mpirun -np 2 ./TpetraCore_BlockCrsPerfTest.exe --num-elements-i=32 --num-elements-j=32 --num-elements-k=32 --num-procs-i=2 --num-procs-j=1 --num-procs-k=1 
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set                                                                                                        
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Kokkos::OpenMP thread_pool_topology[ 1 x 1 x 1 ]
Column = 0  Error norm = 0
==================================================================================================================================

                                              TimeMonitor results over 2 processors

Timer Name                                               MinOverProcs     MeanOverProcs    MaxOverProcs     MeanOverCallCounts
----------------------------------------------------------------------------------------------------------------------------------
0) LocalGraphConstruction                                0.0214 (1)       0.02141 (1)      0.02143 (1)      0.02141 (1)
1) GlobalGraphConstruction                               0.006952 (1)     0.006968 (1)     0.006985 (1)     0.006968 (1)
2) LocalBlockCrsFill                                     0.1801 (1)       0.1804 (1)       0.1807 (1)       0.1804 (1)
3) BlockCrsMatrix FillComplete - currently do nothing    9.537e-07 (1)    1.073e-06 (1)    1.192e-06 (1)    1.073e-06 (1)
4) MultiVectorFill                                       0.003153 (1)     0.003159 (1)     0.003164 (1)     0.003159 (1)
5) BlockCrs Apply                                        4.955 (100)      4.955 (100)      4.955 (100)      0.04955 (100)
6) Conversion from BlockCrs to PointCrs                  0.122 (1)        0.1221 (1)       0.1221 (1)       0.1221 (1)
7) PointCrs Apply                                        5.039 (100)      5.039 (100)      5.039 (100)      0.05039 (100)
8) Export MatrixMarket                                   1.883 (1)        11.92 (1)        21.96 (1)        11.92 (1)
X) Global                                                12.24 (1)        22.28 (1)        32.32 (1)        22.28 (1)
==================================================================================================================================
End Result: TEST PASSED
```

**Comment by ctanis** [comment]
@mhoemmen I'm using a master head from the weekend.. (commit c6810b7889a4e7e8988c0dac587ccc6307acf7f9). 

It's been configured with this script:  
[do-configure.txt](https://github.com/trilinos/Trilinos/files/1891585/do-configure.txt)

I am using openmp and cuda at various times (as well as MPI, obviously).  Attempting to use 1 MPI rank per compute node, and then OpenMP and Cuda intra-node.

**Comment by ctanis** [comment]
@kyungjoo-kim I cannot find this perf test executable or source in my clone of the git repository (Trilinos) or in the build directory.  Do I need to do something special to build this?

**Comment by kyungjoo-kim** [comment]
@ctanis You can find out the code here, 
```
https://github.com/trilinos/Trilinos/tree/master/packages/tpetra/core/example/BlockCrs
```
please set ``EXAMPLES=ON`` in your do-configure.txt. 

This performance test is develpped to identify performance issues and set a baseline performance when we develop a new block crs data structure and an improved apply method. It simply compares block crs and point crs apply methods.

**Comment by mhoemmen** [comment]
To clarify: `Tpetra_ENABLE_EXAMPLES=ON`

**Comment by mhoemmen** [comment]
@ctanis You set `Trilinos_ENABLE_DEBUG:BOOL=ON` in the script.  This enables a bunch of debug checking that will slow down the code quite a bit.

**Comment by mhoemmen** [comment]
@ctanis You normally don't need to set that option at all; just setting `CMAKE_BUILD_TYPE` will automatically set an appropriate default for `Trilinos_ENABLE_DEBUG`.

**Comment by ctanis** [comment]
@mhoemmen I have a non-debug build that shows similar timing.  I'm going to test the Perf script mentioned above tomorrow when I return to work.   did you happen to see the compilation error I got with the BlockVector?  Any ideas about that?  Presumably I can't use a regular MultiVector when NQ>1 (I got multiple issues here.. thanks for your help!)

**Comment by mhoemmen** [comment]
@ctanis wrote:
> did you happen to see the compilation error I got with the BlockVector?

Yes.  The obscureness of the errors suggests to me that you might be trying to use NVCC directly as the compiler, rather than `nvcc_wrapper`.  Please refer to the Tpetra FAQ:

https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/doc/FAQ.txt

**Comment by ctanis** [comment]
@mhoemmen 
> Yes. The obscureness of the errors suggests to me that you might be trying to use NVCC directly as the compiler, rather than nvcc_wrapper. Please refer to the Tpetra FAQ:

I'm definitely using nvcc_wrapper, as without any reference to BlockMultiVector everything compiles and runs on the GPU as appropriate.  The only issue was the parallel performance of the BlockCrsMatrix.apply()

**Comment by mhoemmen** [comment]
@ctanis What version of CUDA are you using?  Also, you may need to supply a `KOKKOS_ARCH` CMake option appropriate to your GPU.  Otherwise Kokkos may not perform well.  See Kokkos documentation for details.

**Comment by ctanis** [comment]
@mhoemmen This is CUDA 8 on a P100.  I had not set KOKKOS_ARCH, so I'm rebuilding Trilinos now.  I'll keep you posted.  In the meantime, @kyungjoo-kim :  my 2 processor version of your performance test is taking about 250 times longer than the 1 processor.  see attached.  I'll try again when the new version finishes building.
[blockmult-2.txt](https://github.com/trilinos/Trilinos/files/1896199/blockmult-2.txt)
[blockmult-1.txt](https://github.com/trilinos/Trilinos/files/1896200/blockmult-1.txt)

**Comment by kyungjoo-kim** [comment]
@ctanis Thank you for running the test.

It says that something clearly goes wrong in the cuda version block crs apply method. My test is limited on the openmp space. We need to dig further about this performance issue and it would take some time. Does this block your work ? or how long can you wait for the improvement ?

**Comment by ctanis** [comment]
@kyungjoo-kim this is pretty important to my work, but as I said I appreciate that the block structures are "experimental".  Is there a way to force the linear solve (and underlying BlockCrs apply) to use OpenMP, but allow me to continue to use Cuda for my matrix assembly?  I have a deadline for a paper on performance portability of my finite element code.

I ran this code on a profiler, and the time is dominated by Kokkos::deep_copy and a Cuda fence()-- I was surprised by this, since I thought Kokkos aggressively opted for UVM.  Maybe I can force UVM space and see if that helps?

**Comment by mhoemmen** [comment]
@ctanis Tpetra has an option to let users control whether MPI tries to talk to CUDA allocations.  

Could you search for `Tpetra_ASSUME_CUDA_AWARE_MPI` in your CMakeCache.txt file (in your Trilinos build directory) and tell me whether it's `ON` or `OFF`?  If it's `ON`, you can set the following environment variable (case sensitive) to turn it off:
```
TPETRA_ASSUME_CUDA_AWARE_MPI=OFF
```

**Comment by kyungjoo-kim** [comment]
@ctanis In my personal opinion, there is no easy way to use openmp space for linear solve while the tpetra object still lives in cuda uvm space. 

The block crs matrix has templarate parameters scalar_type, local_ordinal, global_ordinal, and node_type. So, it might be possible by giving the user provided non default node_type. Probably it does not work in that way. 

The other possible way is to construct two object (one for cuda and the ohter for openmp). Then, you can use cuda object for your FE assembly (I believe that this is your core work for publication) and convert the cuda object to the other one with openmp for linear solve to finish in time.

@mhoemmen Am I saying something crazy ? or feasible ?

**Comment by mhoemmen** [comment]
@kyungjoo-kim That's reasonable.  

@ctanis We've heard that the P100 has some issues with CUDA-aware MPI.  I would recommend setting the environment variable (see my above comment) to force Tpetra to use host memory for MPI communication.

**Comment by ctanis** [comment]
@mhoemmen TPETRA_ASSUME_CUDA_AWARE_MPI is OFF in my CmakeCache.txt.  I'm going to try creating a second non-cuda block matrix for now.  I'm doing a number of other things inefficiently right now, so what's one more!  Please let me know if I can provide any more information.

**Comment by mhoemmen** [comment]
@ctanis Please don't that we don't make any grand promises about BlockCrs performance at the moment.  @kyungjoo-kim and his colleagues are working on OpenMP and CUDA performance actively at the moment.

**Comment by ctanis** [comment]
@mhoemmen Oh, I totally understand!  Is there a good way to accomplish this with the non-Block matrices and vectors?

**Comment by mhoemmen** [comment]
@ctanis You could convert your block matrix to a non-block matrix (plain old `Tpetra::CrsMatrix`) directly.  Just include `Tpetra_Experimental_BlockCrsMatrix_Helpers.hpp` and use the relevant function (look in `Tpetra_Experimental_BlockCrsMatrix_Helpers_decl.hpp` for the declaration and documentation).

**Comment by ctanis** [comment]
@mhoemmen in case I'm missing something, I only see a way to convert from CrsMatrix to BlockCrsMatrix (the other direction) in this file?

**Comment by mhoemmen** [comment]
@ctanis Oops, sorry, that's right.

**Comment by ctanis** [comment]
I've modified all the data structures involved in the linear system to be templated on Kokkos::Compat::SerialWrapperNode ( I tried KokkosOpenMPWrapperNode as well).  In all cases, the multiple process versions perform worse than the single processor.  When I profile, it says that a call to cudaDeviceSynchronize() is dominating the calculation.  

Should this have worked?

    typedef typename Kokkos::Compat::KokkosSerialWrapperNode LSNode;
    //typedef typename ::Kokkos::Compat::KokkosCudaWrapperNode LSNode;

    typedef typename Tpetra::Map<LO, GO, LSNode> LSMapType;
    typedef typename Tpetra::Operator<Scalar, LO, GO, LSNode> OP;
    typedef typename Tpetra::CrsGraph<LO,GO, LSNode> GraphType;
    typedef typename Tpetra::BlockCrsMatrix<Scalar, LO, GO, LSNode> MatrixType;
    typedef typename Tpetra::MultiVector<Scalar,LO,GO, LSNode> MultiVecType;

**Comment by mhoemmen** [comment]
What does "modified the data structures to be templated on SerialWrapperNode" mean?  Does it just mean that you're using Tpetra with the Serial execution space, instead of the Cuda execution space?

**Comment by jhux2** [comment]
>  in case I'm missing something, I only see a way to convert from CrsMatrix to BlockCrsMatrix (the other direction) in this file?

Fwiw, you could write the BlockCRS matrix to file and read it back in as a point CRS matrix, using the Operator write method.

**Comment by ctanis** [comment]
@mhoemmen I used the LSNode typedef in my code snippet as the node_type in the declarations of the various linear algebra objects used in the solve.  I guess I mean that applied the templates in this way -- sorry for the confusing word choice!

**Comment by ctanis** [comment]
Sorry for reviving this old thread, but we finished our paper and I'm back trying to get reasonable Tpetra performance with the block structures.  Specifically I'm trying to use Cuda to build my system and then (for now) use OpenMP to parallelize Tpetra+Belos.  Recall: block matrix-vector multiplication using MPI+Cuda  is extraordinarily slow on our system.

It turns out that simply using a `Kokkos::Compat::KokkosOpenMPWrapperNode` as the node_type parameter for my linear system seems to be making the Matrix-Vector multiplies single-threaded.  I suspect this is related to the comment in the Tpetra FAQ:  "Tpetra enables exactly one Kokkos execution space by default...    - If CUDA is enabled, use Kokkos::Cuda.   - Else, if OpenMP is enabled, use Kokkos::OpenMP.:  

Could it be that if Tpetra is built with Cuda support, OpenMP execution is broken *since* Cuda is enabled?  I tried rebuilding Tpetra without Cuda support, but this is leading to many compilation errors.

I'm also trying constructing my own "node type" that uses OpenMP + CudaUVM, but  again I'm getting errors.  Should something like this work:

`Kokkos::Compat::KokkosDeviceWrapperNode< Kokkos::Device<Kokkos::OpenMP, Kokkos::CudaUVMSpace> >`

Thanks!

**Comment by mhoemmen** [comment]
@jjellio can help with pinning tips.  I do appreciate your patience; we're a bit swamped these days.

**Comment by jjellio** [comment]
The usual is OMP_PLACES=threads, OMP_PROC_BIND=spread, and OMP_NUM_THREADS
The OpenMP things will not help your MPI process mapping though.

Is this with OpenMPI? Or SLURM? (I don't know the mapping for Intel MPI..)

with OpenMPI, you want something like
mpirun --map-by ppr:<Processes_per_Numa>:NUMA:pe=<cpus_per_process>
Or,
mpirun --map-by ppr:<Processes_per_socket>:socket (you can specify pe= if you want).

For example, on my dual socket, dual NUMA domain system:

```
mpirun --map-by ppr:1:NUMA --report-bindings hostname
[s1001943.srn.sandia.gov:96231] MCW rank 1 bound to socket 1[core 14[hwt 0-1]], socket 1[core 15[hwt 0-1]], socket 1[core 16[hwt 0-1]], socket 1[core 17[hwt 0-1]], socket 1[core 18[hwt 0-1]], socket 1[core 19[hwt 0-1]], socket 1[core 20[hwt 0-1]], socket 1[core 21[hwt 0-1]], socket 1[core 22[hwt 0-1]], socket 1[core 23[hwt 0-1]], socket 1[core 24[hwt 0-1]], socket 1[core 25[hwt 0-1]], socket 1[core 26[hwt 0-1]], socket 1[core 27[hwt 0-1]]: [../../../../../../../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB]
[s1001943.srn.sandia.gov:96231] MCW rank 0 bound to socket 0[core 0[hwt 0-1]], socket 0[core 1[hwt 0-1]], socket 0[core 2[hwt 0-1]], socket 0[core 3[hwt 0-1]], socket 0[core 4[hwt 0-1]], socket 0[core 5[hwt 0-1]], socket 0[core 6[hwt 0-1]], socket 0[core 7[hwt 0-1]], socket 0[core 8[hwt 0-1]], socket 0[core 9[hwt 0-1]], socket 0[core 10[hwt 0-1]], socket 0[core 11[hwt 0-1]], socket 0[core 12[hwt 0-1]], socket 0[core 13[hwt 0-1]]: [BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../../../../../../../..]
s1001943.srn.sandia.gov
s1001943.srn.sandia.gov
```
The piece you want to look for, is:
```
[BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../../../../../../../..]
[../../../../../../../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB/BB]
```
That is an ASCII visualization of the CPUMASK applied to each process, i.e., which Cores will the MPI process be allowed to run on.

You may notice, I didn't specify an -np<total_processes>, and OpenMPI simply filled things up following the mapping. You can map-by quite a few things. (check the Manpage). The natural ones are socket or NUMA, if you run on Intel Phi with subnuma clustering, the map-by NUMA feature is very nice.  For OpenMP on traditional things, map-by socket or core may be better.

If your MPI Run doesn't report bindings correctly, it probably wasn't compiled with hwloc correctly... (this happens alot unfortunately).

You can verify the MPI reported bindings like:
```
mpirun --map-by ppr:1:NUMA --display-map bash -c 'echo $OMPI_COMM_WORLD_RANK $(cat /proc/$$/status | grep Cpus_allowed_list)'

0 Cpus_allowed_list: 0-13,28-41
1 Cpus_allowed_list: 14-27,42-55
```

That's some ugly (probably not safe) bash execution, but what it will do is use mpirun to spawn off bash processes. The processes then inspect their own /proc/PID/status, and report the CPU list of themselves.... there are other ways to do it.

There are also a number of OMPI_* env variables, can you see them by `mpirun -np 1 env`, or maybe in the MPI man pages (somewhere buried most likely).

Compare the output of the CPUMASK stuff to tools like:
```
numactl -H

hwloc-ls
```

If you are on a desktop with X11, I highly recommend, 
```
lstopo
```
(Part of Hwloc), it will print a nice picture showing your machines topology.

**Comment by ctanis** [comment]
@jjellio Thanks for this.. I'm seeing something a little different here.  If I don't use a --bind-to clause, it binds each rank to a single core.  Have you seen this?  Am I going to have ill effects from using bind-to?
I'm going to test the performance myself, but I'm having to rebuild Trilinos with a different version of MPI (You were right.. I was using one with a broken hwloc)

```
os-hn:~$ mpirun --map-by ppr:1:numa -machinefile /scr/ctanis/kkfe/machine1.txt  bash -c 'echo $OMPI_COMM_WORLD_RANK $(cat /proc/$$/status | grep Cpus_allowed_list)'
0 Cpus_allowed_list: 0
1 Cpus_allowed_list: 1
os-hn:~$ mpirun --map-by ppr:1:numa --bind-to numa -machinefile /scr/ctanis/kkfe/machine1.txt  bash -c 'echo $OMPI_COMM_WORLD_RANK $(cat /proc/$$/status | grep Cpus_allowed_list)'
0 Cpus_allowed_list: 0,2,4,6,8,10,12,14,16,18,20,22,24,26
1 Cpus_allowed_list: 1,3,5,7,9,11,13,15,17,19,21,23,25,27
```

**Comment by jjellio** [comment]
Yes, you will have issues.

You don't want your process masks to look like the bind-to NUMA one. That mapping is giving you a weird looking mapping of CPUids. Normally, the cpuids will follow the pattern: 
core0_ht0, core1_ht0, ....,coreN_ht0, core0_ht1, ...

I can't think of a reason you would want MPI processes sharing a core.

what is the output of `hwloc-ls` on a single node.  That will tell me what the mapping should be. (or `numactl -H`)

**Comment by ctanis** [comment]
On this node, the cpu ids are staggered on each NUMA node:
```
numactl -H
available: 2 nodes (0-1)
node 0 cpus: 0 2 4 6 8 10 12 14 16 18 20 22 24 26
node 0 size: 65440 MB
node 0 free: 50317 MB
node 1 cpus: 1 3 5 7 9 11 13 15 17 19 21 23 25 27
node 1 size: 65536 MB
node 1 free: 59180 MB
node distances:
node   0   1 
  0:  10  21 
  1:  21  10 
```

**Comment by jjellio** [comment]
That is very odd. Usually an entire socket will get lumped into a NUMA region (since the CPU effectively has the memory controller), and cpuids are usually numbered by socket. Some AMD chips pack two CPUs (dies) onto a single socket, and the OS sees two memory regions.

I'd be skeptical about numactl on that machine. It could correct.  Either way,

set OMP_DISPLAY_ENV=verbose, (and whatever you want for threads and places, proc-bind=spread)

I'd run two processes on that node, and you will get two burps of OMP output at the start. You want to see what OMP_PLACES looks like in that output. With GNU OpenMP, PLACES will get expanded to sets of CPU ids e.g., {0,10}, {1,11}, etc... if you have PLACES=cores, and each core has two hardware threads.

**Comment by ctanis** [comment]
It looks like I'm distributing threads properly.. I can get 100% CPU utilization across all cores with a dummy program that grinds away with a compute-intensive Kokkos::parallel_for

With the same launch settings, this Tpetra solve reverts to the described situation where one core is pegged, and the others do nothing.   As an experimetn, I set OMP_WAIT_POLICY to active, and all the cores are firing at 99% .   This, of course, does not speed up the solution, but is it a clue perhaps about the slow solve?

Thanks

**Comment by mhoemmen** [comment]
@ctanis I should ask the obvious question: What's the Node type?  Are you using Tpetra's default Node type?  What is it?  (Try printing `typeid(Tpetra::Map<>::device_type::execution_space).name()`, or `typeid(Node::device_type::execution_space).name()` if you are using a different Node type.

**Comment by mhoemmen** [comment]
@ctanis You could also try some simpler Tpetra benchmarks, like the Vector benchmark:

https://github.com/trilinos/Trilinos/blob/develop/packages/tpetra/core/example/advanced/Benchmarks/vector.cpp

This should illustrate whether threads are taking any effect, possibly after adjusting the command-line arguments a bit.

**Comment by ctanis** [comment]
@jjellio I suggested it as a hint because when I am not careful with the map-to and bind settings, `OMP_WAIT_POLICY=active` only lights up one core.  I can see all 28 cores peg now, so that suggests that I'm at least launching the MPI job properly.  

I am even seeing reasonable thread scaling with a pure OpenMP build now, but when I build a hybrid that uses `Kokkos::Cuda` for building my linear system and then a `Kokkos::OpenMP` to solve it, the thread bindings seem to be affected by the presence of Cuda.  To answer your question, @mhoemmen , in this case the `execution_space.name()`'s are `N6Kokkos4CudaE` and `N6Kokkos6OpenMPE` respectively (using `CudaUVMSpace` as a memory space).  Could there be a problem using CudaUVMSpace from OpenMP inside Tpetra?  I am doing this in my own code and it behaves as expected.

I'm going to try to see how well this works with the traditional (non block) Tpetra structures.

**Comment by jjellio** [comment]
I don't think this is a binding issue.

I've used code like the following to inspect the objects I get at runtime. You can dump your  solvers/datatypes into ` Teuchos::demangleName(typeid(     ).name())`  and get a feel for what types you have hanging around.  My guess is you have a mismatch. But with mixed Execution spaces I'm not really sure.

```
  // These are the template parameters used to instantiate some objects
  ParameterList configPL ("Configuration");
  ParameterList& ptypePL = configPL.sublist("Primitive Types", false, "Primitive Type Information");
  ptypePL.set("Scalar", Teuchos::demangleName(typeid(SC).name()));
  ptypePL.set("LocalOrdinal", Teuchos::demangleName(typeid(LO).name()));
  ptypePL.set("GlobalOrdinal", Teuchos::demangleName(typeid(GO).name()));
  ptypePL.set("Node", Teuchos::demangleName(typeid(NO).name()));

  // these are types, e.g. MultiVector<SC,LO,GO,NO> 
  ParameterList& typePL = configPL.sublist("Type Information", false, "Type Information");
  typePL.set("MV", Teuchos::demangleName(typeid(MV).name()));
  typePL.set("Map", Teuchos::demangleName(typeid(map_type).name()));

configPL.print(<some ostream>)
```

**Comment by jjellio** [comment]
Another debug option:
If you want to eliminate MPI,

(copy one list of CPUs from a numa domain)
```
cpu_list="0 2 4 6 8 10 12 14 16 18 20 22 24 26"
numa_domain=0
export OMP_NUM_THREADS=$(echo "${cpu_list}" | wc -w)
... set places, proc bind
numactl -C $(echo "${cpu_list}" | tr ' ' ',') --membind=${numa_domain} ./your_binary --your-args ...
```

That will bind your process to the cpus in the list, and memory bind your process to specific NUMA domain.   I doubt this will make a difference. But it could eliminate MPI and NUMA effects as possible problems.


Also, I assume you are setting:
export CUDA_LAUNCH_BLOCKING=1
export CUDA_MANAGED_FORCE_DEVICE_ALLOC=1

Kokkos will whine if you don't.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> Could there be a problem using CudaUVMSpace from OpenMP inside Tpetra?

CudaUVMSpace allocations live on the GPU.  If you try to access them from multiple CPU threads, you probably won't get the memory bandwidth that you would normally get with Kokkos CPU allocations.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> but when I build a hybrid that uses Kokkos::Cuda for building my linear system and then a Kokkos::OpenMP to solve it

This only makes sense if using a host-only solver, say domain decomposition with a sparse direct subdomain solver.

**Comment by ctanis** [comment]
@mhoemmen 

> CudaUVMSpace allocations live on the GPU

I thought they lived in both?  In fact I just did an experiment that used OpenMP+HostSpace kernel and it worked on CudaUVMSpace Views without any warnings or problems.

I'm highly suspicious now of the distributed graph / maps I'm using for my matrix.  Recall that the performance is amazingly bad when MPI is involved.  Do you have any tips for verifying that my distributed objects are optimal?  I build a final graph from my own compressed row indices and use that for the BlockCsrMatrix.  The number of nonzeros on each row is fixed, etc.

> This only makes sense if using a host-only solver, say domain decomposition with a sparse direct subdomain solver.

Yeah the hybrid approach here is only because the MPI performance of the linear solve is so terrible when Cuda is involved.

**Comment by ctanis** [comment]
@mhoemmen 

> How do you build the graph's row Map?

I'm glad you asked this, as I've been revisiting this part of my code today.  I am  hoping that part of my problem is that I've misunderstood something and used a naive column map in my graph as well.
The row map is a contiguous map based on the range of global ids owned by each rank.  I used a column map to map local indices to phantom nodes owned by other processes, but I think this is causing a lot more communication than necessary during the matrix-vector multiply.  

Logically I have everything that I need on each MPI rank, but the matrix needs to know about vector data on other elements.  How should this be done?  I couldn't find a good example of this with an unstructured problem.

> CudaUVMSpace allocations live on the GPU, but are accessible from host.
How does this work with GPU page faults where UVM can be larger than GPU memory?  Your point on simultaneous access is well-taken!

Thanks for all your help.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> I used a column map to map local indices to phantom nodes owned by other processes, but I think this is causing a lot more communication than necessary during the matrix-vector multiply.

For a first step, try creating the graph without a column Map, using global indices.  Let the graph's fillComplete method create the column Map automatically.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> How does this work with GPU page faults where UVM can be larger than GPU memory?

I'm not quite sure how that works, but it's certainly possible to run out of GPU memory.  I've never tried the experiment of allocating more than GPU capacity and accessing it sparsely.  I strongly dislike overcommit because it defers bug detection until first access, but some applications find it useful.

**Comment by ctanis** [comment]
@mhoemmen 
> For a first step, try creating the graph without a column Map, using global indices. Let the graph's fillComplete method create the column Map automatically.

This does not seem to help, but I do have some rather unusual results to share.  I'd like to preface this by saying my code's output is correct.  This suggests that my graph/matrix is logically valid, if not optimal. 

I am timing a matrix-vector multiply (matrixrcp->apply(...)) here, where my matrix is the BlockCsrMatrix as discussed.  In each case, I'm setting OMP_NUM_THREADS=1 to focus on MPI overhead.
Running with 1 or 2 MPI processes:

- Trilinos serial-only build
NP=1, time=.28 seconds
NP=2, time=.15 seconds

- Trilinos built with Cuda+OpenMp support, but   Tpetra is instantiated on `::Kokkos::Compat::KokkosOpenMPWrapperNode`  (the GPU is observed to be idle during this part of the test)
NP=1, time=.275 seconds
NP=2, time=14.9 seconds

The memory space for the Tpetra node is printing as `N6Kokkos9HostSpaceE`

It appears that the mere presence of Cuda is nerfing my MPI performance, independent of any inter-node parallelism.

**Comment by mhoemmen** [comment]
@ctanis That's weird... Did you try this with a non-block (CrsMatrix) sparse matrix-vector multiply?  We have some benchmarks you can run.

**Comment by ctanis** [comment]
@mhoemmen 

> That's weird... Did you try this with a non-block (CrsMatrix) sparse matrix-vector multiply? We have some benchmarks you can run.

Just got back from a long trip, and this is at the top of my todo list.  thanks, and I'll keep you posted!

**Comment by ctanis** [comment]
@mhoemmen it looks like the non-block (CrsMatrix) performs as expected..  Can you point me at the benchmarks you mentioned?

**Comment by mhoemmen** [comment]
@ctanis Try `Trilinos/tpetra/core/test/PerformanceCGSolve/`.  It builds if you enable Tpetra's tests (CMake option `Tpetra_ENABLE_TESTS:BOOL=ON`).  The executable reads in the CrsMatrix as a Matrix Market file, then benchmarks the method of conjugate gradients (so, sparse matrix-vector multiply, plus dot products, norms, and vector updates).

**Comment by mhoemmen** [comment]
That setting is a work-around.  @ambrad could also help comment on this as I think he implemented the work-around.  `doImport` just does the communication for the sparse mat-vec.

I'm swamped today; I just don't want you to think I'm ignoring you :-) .

**Comment by ctanis** [comment]
@mhoemmen I appreciate all your help!

FWIW, I built Trilinos with HAVE_TPETRA_BCRS_DO_POINT_IMPORT and the MPI+Cuda performance is far improved.  I'm not sure it scales in MPI, but it no longer explodes on me.  I'll have more information when I fix some of my own bugs that are making it difficult to try larger jobs.  I just wanted you to know that this option appeared to help performance.

**Comment by ambrad** [comment]
Sorry: just saw this. Yes, setting `-D Tpetra_BCRS_Point_Import=ON` in your Trilinos configuration will improve the block CRS matvec, as the MPI comm will then funnel through a code path that is intended to run reasonably well on GPU.

**Comment by mhoemmen** [comment]
Thanks @ambrad ! :-D

The whole MPI + CUDA story is a bit of a mess at the moment.  It might _work_, and many code paths use the GPU to pack and unpack messages.  However, MPI send and receive directly to/from GPU memory, while it works, seems to be slow.  We have a lot more work to do.

**Comment by mhoemmen** [comment]
Closing -- thanks all!

**Comment by ctanis** [comment]
sorry to rehash this, but block sparse matrix performance with Cuda+MPI seems to have reverted to nightmarish performance (still building with Tpetra_BCRS_Point_Import=ON) ... any ideas what could have happened?

**Comment by kddevin** [comment]
@kyungjoo-kim @trilinos/tpetra

**Comment by kyungjoo-kim** [comment]
@kddevin  

There was some changes in the native Tpetra Block Crs importer (now importer pack data from device). However, with ``Tpetra_BCRS_Point_Import=ON``, the MPI comm goes through point CRS matrix infra structures. If @ctanis  sees slowness compared to some previous version, this might not be an issue on the block crs but the crs matrix importer.   Do we have any report from applications regarding to cuda+mpi slowness while proceeding deprecation ?

**Comment by ctanis** [comment]
@kyungjoo-kim Should I try rebuilding Tpetra without Tpetra_BCRS_Point_Import turned on?
Another thing I've noticed is that it appears now that by default Tpetra assumes that cuda-aware MPI is enabled -- could this be related?

**Comment by kyungjoo-kim** [comment]
@ctanis The new BCRS importer is part of refactoring work for Tpetra. If the slowness comes from the importer, then it won't change the performance (if the performance slowness is from communication). 

We had some troubles in cuda MPIs. I would say that it could be related. For the apple to apple comparison, you may want to test without cuda mpi if previously the option is not on. 

Also, as we do not keep track of the performance of BCRS, we cannot really comment whether the performance degrade is Tpetra problem or the mpi comm at this moment. I will get back to you after evaluating with different configurations.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> Another thing I've noticed is that it appears now that by default Tpetra assumes that cuda-aware MPI is enabled -- could this be related?

If you have OpenMPI, Tpetra asks your OpenMPI installation at configure time whether MPI is CUDA aware.  However, this is just a default value.  You can control this behavior at run time by setting the `TPETRA_ASSUME_CUDA_AWARE_MPI` environment variable (to 1 for "yes, do assume," or to 0 for "do not assume that MPI is CUDA aware").

**Comment by ctanis** [comment]
I'm not sure what I was doing wrong, but I rebulit Trilinos using a newer version of openmpi (with better MPI+cuda support) and configured it using the Tpetra_BCRS_Point_Import:BOOL=ON  and I am getting decent performance again with my sparse block operations.  If I can figure out any consistent patterns here, I'll let you know.

**Comment by mhoemmen** [comment]
@ctanis We've definitely seen newer versions of OpenMPI do better.  @jjellio could say more.

**Comment by jjellio** [comment]
@ctanis  For Cuda + MPI, it has been my experience that you need OpenMPI above 2, and probably at or above 4.0.
The kicker, is that your Cuda+MPI performance is currently dominated by the underlying transport + network card you use.  On the advanced machines, we use UCX (from mellanox) or PAMI (from IBM).  To get good performance, the transport layer in OpenMPI needs to be able to do direct memory access (DMA) from Device to NIC. If DMA isn't working, you end up with MPI doing copies from device to Host, and then normal MPI from host to host, followed by a host to device copy.

If you can post the output of:
```
mpirun \
<Your regular MPI run options, -np --map-by, etc.. > \
-mca btl_base_verbose 99 \
-mca pml_base_verbose 99 \
-mca mtl_base_verbose 99 \
-mca osc_base_verbose 99 \
<some_mpi_program>
```

Then we can see what transports your cluster supports. My guess, is that your have either TCP or OpenIB (infiniband) and neither of those are Cuda 'aware' ...  

(Locate the Tree of Woe and continue reading)
Trilinos' Cuda aware checks simply make a best effort....

(start crying now)

Unfortunately, OpenMPI doesn't tell the whole truth of Cuda+MPI with 'cuda aware = true'.  When OpenMPI says it is Cuda aware, it simply means it won't segfault if you give MPI a pointer to Device memory. It isn't telling you whether that will be handled performantly though.

(begin to wail and gnash teeth)

What OpenMPI isn't reporting is which transports Cuda+MPI will actually use.  We've had issues where we have the *correct transport, compiled with cuda support* and still it follows the wrong code path in OpenMPI because the internal (openmpi) configuration gets botched.

The flags above will report the transport selection process (basically verbose MPI_Init)

**Comment by ctanis** [comment]
I've done some more testing, and it seems that I'm having some pretty fundamental problems with Kokkos performance in Tpetra.    I'm running a small enough test that I dont' need to use MPI, so this is just looking at Kokkos performance, in solving a sparse linear system with 890k nodes and 12.8M nonzeros:

Kokkos::Cuda:  21 seconds
Kokkos::OpenMP (28 threads):  20 seconds
28 MPI processes, Kokkos::OpenMP with 1 thread each:  1.7 seconds

I can provide more detailed info here, as I know this is not much to go by.  I'm not entirely sure how to profile this, but watching top in the OMP cases, it's clear that most cores are idling most of the time.  This suggests some memory conflicts, but I'm using Tpetra defaults.

matrix assembly using Kokkos for loops perform very well.

**Comment by jjellio** [comment]
Can you share your `mpirun` line and `numactl -H`.

**Comment by ctanis** [comment]
```
ts01:report0719$ numactl -H
available: 2 nodes (0-1)
node 0 cpus: 0 2 4 6 8 10 12 14 16 18 20 22 24 26
node 0 size: 65440 MB
node 0 free: 51342 MB
node 1 cpus: 1 3 5 7 9 11 13 15 17 19 21 23 25 27
node 1 size: 65536 MB
node 1 free: 60214 MB
node distances:
node   0   1 
  0:  10  21 
  1:  21  10 
```

 ```mpirun  -machinefile machines.txt  --map-by ppr:1:node --bind-to none -np 1 -x OMP_PROC_BIND=spread -x OMP_PLACES=threads -x CUDA_LAUNCH_BLOCKING -x CUDA_MANAGED_FORCE_DEVICE_ALLOC -x LD_LIBRARY_PATH -x OMP_NUM_THREADS=28```

I've found that without "--bind-to none", OMP_DISPLAY_ENV (and subsequent behavior) appears to be unaware of cores.  Again, my Kokkos::parallel_for loops get good performance across all cores with this configuration, but the Tpetra stuff falters.

**Comment by jjellio** [comment]
Can you try:
`
# for OMP with processes w/14 threads per proc
# you can probably skip OMP_NUM_THREADS, since it will default to the number of 'processors'
# which, with proper binding will be $threads_per_socket

threads_per_socket=14;
procs_per_socket=$((14/${threads_per_socket}))
num_procs=$((${procs_per_socket}*2))

mpirun \
-x OMP_PROC_BIND=spread -x OMP_PLACES=cores \
-x CUDA_LAUNCH_BLOCKING \
-x CUDA_MANAGED_FORCE_DEVICE_ALLOC \
-x LD_LIBRARY_PATH \
--map-by ppr:${procs_per_socket}:socket:pe=${threads_per_socket} \
--bind-to core 
# you can actually skip -np, since map-by is going to evenly populate the node
-np ${num_procs}
```

Effectively, you want something like: `--map-by ppr:14:socket:pe=1 --bind-to core` for flat MPI, and then you trade `ppr:14` with `pe=1` to trade threads for processes.  I usually test with

`mpirun <flags> --report-bindings numactl -s`

That will show the MPI process binding map (one row per process). Then the program `numactl -s` will showing the process binding it actually has (should be the same as a corresponding row of the OpenMPI process binding report)

You probably won't get great performance with a single process since you have two sockets. 

See if the above helps with the OpenMP perf... I suspect you are getting OpenMP threads in a 'team' that are shared across a socket.

**Comment by nmhamster** [comment]
What hardware is this? The thread layout is weird.

The `--reporting-bindings` that James asked for is important. I am wondering if the hwloc is just blowing up here?

**Comment by ctanis** [comment]
Not sure this is adequate hardware info for you, but the CPU is a "Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz".  I'm not sure why the numa nodes contain alternating core numbers.  Here's the result of James's mpirun configuration, once using numactl -H, and once with an excerpt from my logs with OMP_DISPLAY_ENV=true

It does look like hwloc is getting things wrong, but then OpenMP is reporting the right places.  
```
os-hn:report0719$ mpirun --machinefile machines.txt -x OMP_PROC_BIND=spread -x OMP_PLACES=cores -x CUDA_LAUNCH_BLOCKING -x CUDA_MANAGED_FORCE_DEVICE_ALLOC -x LD_LIBRARY_PATH -x OMP_DISPLAY_ENV=true --map-by ppr:1:numa:pe=14 --bind-to core -np 2 --report-bindings   numactl -H
available: 2 nodes (0-1)
node 0 cpus: 0 2 4 6 8 10 12 14 16 18 20 22 24 26
node 0 size: 65440 MB
node 0 free: 51327 MB
node 1 cpus: 1 3 5 7 9 11 13 15 17 19 21 23 25 27
node 1 size: 65536 MB
node 1 free: 60213 MB
node distances:
node   0   1 
  0:  10  21 
  1:  21  10 
[ts01:98244] MCW rank 0 bound to socket 0[core 0[hwt 0]], socket 0[core 1[hwt 0]], socket 0[core 2[hwt 0]], socket 0[core 3[hwt 0]], socket 0[core 4[hwt 0]], socket 0[core 5[hwt 0]], socket 0[core 6[hwt 0]], socket 0[core 7[hwt 0]], socket 0[core 8[hwt 0]], socket 0[core 9[hwt 0]], socket 0[core 10[hwt 0]], socket 0[core 11[hwt 0]], socket 0[core 12[hwt 0]], socket 0[core 13[hwt 0]]: [B/B/B/B/B/B/B/B/B/B/B/B/B/B][./././././././././././././.]
[ts01:98244] MCW rank 1 bound to socket 1[core 14[hwt 0]], socket 1[core 15[hwt 0]], socket 1[core 16[hwt 0]], socket 1[core 17[hwt 0]], socket 1[core 18[hwt 0]], socket 1[core 19[hwt 0]], socket 1[core 20[hwt 0]], socket 1[core 21[hwt 0]], socket 1[core 22[hwt 0]], socket 1[core 23[hwt 0]], socket 1[core 24[hwt 0]], socket 1[core 25[hwt 0]], socket 1[core 26[hwt 0]], socket 1[core 27[hwt 0]]: [./././././././././././././.][B/B/B/B/B/B/B/B/B/B/B/B/B/B]
available: 2 nodes (0-1)
node 0 cpus: 0 2 4 6 8 10 12 14 16 18 20 22 24 26
node 0 size: 65440 MB
node 0 free: 51327 MB
node 1 cpus: 1 3 5 7 9 11 13 15 17 19 21 23 25 27
node 1 size: 65536 MB
node 1 free: 60215 MB
node distances:
node   0   1 
  0:  10  21 
  1:  21  10 

```
and then...
```
OPENMP DISPLAY ENVIRONMENT BEGIN
  _OPENMP = '201307'
  OMP_DYNAMIC = 'FALSE'
  OMP_NESTED = 'FALSE'
  OMP_NUM_THREADS = '14'
  OMP_SCHEDULE = 'DYNAMIC'
  OMP_PROC_BIND = 'SPREAD'
  OMP_PLACES = '{0},{2},{4},{6},{8},{10},{12},{14},{16},{18},{20},{22},{24},{26}'
  OMP_STACKSIZE = '1882513440334153457'
  OMP_WAIT_POLICY = 'PASSIVE'
  OMP_THREAD_LIMIT = '4294967295'
  OMP_MAX_ACTIVE_LEVELS = '2147483647'
  OMP_CANCELLATION = 'FALSE'
  OMP_DEFAULT_DEVICE = '0'
OPENMP DISPLAY ENVIRONMENT END

OPENMP DISPLAY ENVIRONMENT BEGIN
  _OPENMP = '201307'
  OMP_DYNAMIC = 'FALSE'
  OMP_NESTED = 'FALSE'
  OMP_NUM_THREADS = '14'
  OMP_SCHEDULE = 'DYNAMIC'
  OMP_PROC_BIND = 'SPREAD'
  OMP_PLACES = '{1},{3},{5},{7},{9},{11},{13},{15},{17},{19},{21},{23},{25},{27}'
  OMP_STACKSIZE = '2603089380713432817'
  OMP_WAIT_POLICY = 'PASSIVE'
  OMP_THREAD_LIMIT = '4294967295'
  OMP_MAX_ACTIVE_LEVELS = '2147483647'
  OMP_CANCELLATION = 'FALSE'
  OMP_DEFAULT_DEVICE = '0'
OPENMP DISPLAY ENVIRONMENT END
```

**Comment by jjellio** [comment]
That looks good:

```
MCW rank 0 bound to socket 0 ...  [B/B/B/B/B/B/B/B/B/B/B/B/B/B][./././././././././././././.]
MCW rank 1 bound to socket 1 ...  [./././././././././././././.][B/B/B/B/B/B/B/B/B/B/B/B/B/B]
```
The braces indicate a socket, and this shows your MPI proc is isolate to a single socket each.
Moreover, OpenMP sees the proper cores (1,3,5....) and (0,2,4...)

How does this impact your OpenMP performance?

**Comment by mhoemmen** [comment]
@ctanis How are you solving the linear system?

**Comment by ctanis** [comment]
@jjellio Does it concern you that report-bindings indicates that MPI thinks the cores are numbered differently?  I agree that it looks like OpenMP knows what threads belong to which cores, but in watching top there's only one thread per MPI rank that's doing any work.  Since this configuration doubles the number of MPI processes, it is performing twice as fast.  I can take this all the way down to 28 MPI ranks , each with one thread, and it is really fast.

@mhoemmen I'm solving with Belos operating on the vector views from the block structures (BlockCRSMatrix, BlockVector).  This is using an Ifpack2 RBILUK preconditioner.

**Comment by jjellio** [comment]
The OpenMPI / OpenMP stuff looks correct. OpenMPI is binding the process to a socket, and OpenMP is reporting the correct CPUIDs.

If the map looks right, and the OpenMP verbose looks right, then it most likely is correct.

With Kokkos + GPU, you do need to set '--kokkos-ndevices=2' or however many GPUs you have per node.  Kokkos will then bind to unique GPUs... without that, it will lump them all on the same device.

For OpenMP, what you have should be correct.... it sounds like there is a just a big serial region.

You can also clone/build
https://github.com/kokkos/kokkos-tools

Usually, https://github.com/kokkos/kokkos-tools/tree/master/src/tools/space-time-stack

Then run with
`-x KOKKOS_PROFILE_LIBRARY=/path/to/src/kokkos-tools/src/tools/space-time-stack/kp_space_time_stack.so`

That will enable the Kokkos Profiler, and it will dump a info about how much time is getting spent inside kokkos (as well as what time is not spent in kokkos) - it does add overhead, but you can generally pick out the oddball timer that way.

**Comment by mhoemmen** [comment]
> I'm solving with Belos operating on the vector views from the block structures (BlockCRSMatrix, BlockVector). This is using an Ifpack2 RBILUK preconditioner.

Would you consider trying the "performance CG solve" example in `tpetra/core/test/`?  This could be a block crs issue.

**Comment by ctanis** [comment]
@mhoemmen I can run whatever you'd like, but it's not clear how to best run this test on a block matrix.
I do have some more information.  It does look like the preconditioner is taking longer than it should.  See this report from Belos:

```
Timer Name                                        MinOverProcs    MeanOverProcs    MaxOverProcs    MeanOverCallCounts    
-------------------------------------------------------------------------------------------------------------------------
Belos: ICGS[2]: Ortho (Inner Product)             1.064 (36)      1.084 (36)       1.104 (36)      0.03012 (36)          
Belos: ICGS[2]: Ortho (Norm)                      0.05321 (19)    0.06608 (19)     0.07894 (19)    0.003478 (19)         
Belos: ICGS[2]: Ortho (Update)                    1.649 (36)      1.674 (36)       1.698 (36)      0.04649 (36)          
Belos: ICGS[2]: Orthogonalization                 2.826 (19)      2.834 (19)       2.842 (19)      0.1492 (19)           
Belos: Operation Op*x                             0.668 (20)      0.859 (20)       1.05 (20)       0.04295 (20)          
Belos: Operation Prec*x                           6.882 (19)      7.003 (19)       7.124 (19)      0.3686 (19)           
Belos: PseudoBlockGmresSolMgr total solve time    10.7 (1)        10.71 (1)        10.71 (1)       10.71 (1)             
Ifpack2::Preconditioner::compute                  4.363 (1)       4.442 (1)        4.521 (1)       4.442 (1)             
Ifpack2::Preconditioner::condest                  0 (0)           0 (0)            0 (0)           0 (0)                 
Ifpack2::Preconditioner::initialize               0.2999 (1)      0.3026 (1)       0.3053 (1)      0.3026 (1)     
```

I'm surprised because I expected this preconditioner operator to look a lot like the original matrix (templated on the same node_type) but it does look like it's not using Kokkos properly.  I followed this example code to make the preconditioner, switching precondType to RBILUK and setting fill-level to 0:

https://github.com/trilinos/trilinos_tutorial/wiki/Ifpack2CreatePreconditioner

TpetraMatrixType is a Tpetra::BlockCRSMatrix

**Comment by ctanis** [comment]
I commented out the use of the preconditioner and it looks like the block matrix operations continue to only use 1 core per MPI rank,so it's not just Ifpack2

**Comment by jjellio** [comment]
I'd build the Kokkos profiler (message above) and then run with this ENV set `-x KOKKOS_PROFILE_LIBRARY=/path/to/src/kokkos-tools/src/tools/space-time-stack/kp_space_time_stack.so`

That will show you the times broken down into Kokkos regions. You can see inside the matrix apply that way.

**Comment by mhoemmen** [comment]
@ctanis A lot of the block crs stuff isn't necessarily using Kokkos properly.  We've optimized very specific code paths there for specific reasons.  I'm working on parallelizing more things, since we need them for other reasons.

**Comment by ctanis** [comment]
@mhoemmen that is good to know.  block sparse matrix performance is important to us (building a multiphysics FE solver), and I'd like to try to help.  I'm primarily trying to distinguish between a problem with the code and user error.

**Comment by mhoemmen** [comment]
@ctanis Thanks so much for your help!  Do you still see these issues when you omit the preconditioner completely?  Also, what kernels are most important to you?

**Comment by ctanis** [comment]
@mhoemmen  We are struggling to compare apples to apples here because when we turn off the preconditioner the matrix takes too long to solve.  I'm attaching the output of hp_time_stack after one iteration with the preconditioner, hoping you can help me make sense of it!  There's a Cuda file and two OpenMP files with different thread counts.  These are running entirely on one node so that  MPI is not part of the timing.  It looks like the time it takes to compute and apply Prec*x is independent of the Kokkos configuration.  

[cuda.txt](https://github.com/trilinos/Trilinos/files/3393999/cuda.txt)
[omp_28threads.txt](https://github.com/trilinos/Trilinos/files/3393997/omp_28threads.txt)
[omp_2threads.txt](https://github.com/trilinos/Trilinos/files/3393998/omp_2threads.txt)


Our primary need right now is to solve the linear system, and depending on the physics our blocks could be  almost any size.  For now 5x5 blocks are being used.

Let me know if there's anything else I can provide.

Thanks

**Comment by mhoemmen** [comment]
@ctanis Our BlockCrsMatrix triangular solves don't use threads at all right now.  I'm working on that :-) .  Thanks for the example!

**Comment by jjellio** [comment]
Why does your Cuda log have all of the dual view tracking memory allocations... I thought that stuff got nuked?

**Comment by ctanis** [comment]
@mhoemmen well, that'd do it!  I'm not asking for a commitment here, but is this an active project?

@jjellio not sure what you mean..

**Comment by jjellio** [comment]
Look at the end of your Cuda log for '0.5% DualView::modified_flags
  0.5% DualView::modified_flags'.  I am not accustomed to seeing those anymore - perhaps there is something slightly different in how I build - or I simply haven't noticed them?

**Comment by ctanis** [comment]
I call sync and modify on DualViews.  Is this no longer necessary?

**Comment by jjellio** [comment]
I am just curious why I don't remember seeing those memory allocations recently.

With UVM, cuda > 8.0, and Launch blocking on, there is no need to sync anything - but I am sure others will disagree with me on that. Using the API is probably the right thing to do, because at some point it would be nice to remove CUDA_LAUNCH_BLOCKING=1 from the required runtime ENV.

Glancing at some recent runs, I don't have those memory allocations - I am just curious if they should/should not be there.

**Comment by mhoemmen** [comment]
@jjellio wrote:
> With UVM, cuda > 8.0, and Launch blocking on, there is no need to sync anything

What happens if a kernel is running on device concurrently with host access?  Does that "just work" now?

**Comment by ctanis** [comment]
I don't use UVM outside of the Tpetra stuff (where it's required, right?) in my code, because I would rather crash than swap on the device.

**Comment by kyungjoo-kim** [comment]
I just want to let you know that I am going to evalute block crs infrastructure and will put the result in the benchmark directory as well as diagnosing the all the things discussed here.

**Comment by mhoemmen** [comment]
@ctanis wrote:
> I don't use UVM outside of the Tpetra stuff (where it's required, right?)

That's correct.  The "swap on the device" thing is not so bad actually.  @trilinos/muelu developers have done studies showing that after setup, the solves don't swap from the GPU.

**Comment by ctanis** [comment]
@kyungjoo-kim excellent!  this benchmarking directory is in the trilinos distribution?

@mhoemmen that's good to know, but what if the system doesn't fit on the GPU to begin with -- when is that detectable?

**Comment by kyungjoo-kim** [comment]
Yeah... you once ran the benchmark code before ``tpetra/core/example/BlockCrs``.

**Comment by jjellio** [comment]
@mhoemmen With Launch blocking on, it would impossible for you to have in-process concurrency active while also running the kernel, unless you have written your own background threads. For Vanilla Tpetra I don't see the point.

I do think it is good to use the API, because baking 'cuda-isms' into the library is just bad.

@ctanis  I've seen very little issues with UVM. If your data primarily lives on the device, but you'd also like to touch it on host - it works quite well. I think the catch is to make sure to touch the memory in different execution spaces judiciously.  In some small tests, I've seen no real difference between the UVM and CudaSpace - if Tpetra is touching memory in a different ExecSpace, we should find and fix that.

UVM semantics seem to say that you should never try to modify memory in both places concurrently. If launch blocking was off, then you would need to ensure that any kernel launched that touched the memory was finished (hence a sync). Likewise you would need to make sure not have concurrent threads that could modify the memory while a kernel spins up. I am guessing it would just thrash between host and device. Also, while Kokkos/Trilinos says that you need to CUDA_MANAGED_FORCE_DEVICE_ALLOC=1, from Cuda 8 onwards, UVM became a page-level kernel module, and so memory is never actually allocated on device or host... instead, memory is simply 'allocated' in a purely virtual sense, and then wherever you read/write the memory is where it becomes allocated (faulted technically).  This is interesting, because it means you can have different 'initialization' behavior. That is, on the Host, memory pages may be promised to be faulted with zero's, while on a GPU I don't think Nvidia they make a promise.

**Comment by ctanis** [comment]
@jjellio I may try switching to UVM at some point in my code, but the current setup has discrete waves of Host and then Device processing, and the explicit syncs feel appropriate.  As I said earlier, I'd rather this part just fail if for some reason the allocation is too large for the GPU

Now that I'm thinking about it, Cuda might now allow you to malloc more memory than the device physically contains even when not using UVM.  Do you know?

**Comment by jjellio** [comment]
I'd try to reason out weird performance more, before assuming UVM is an issue though. The problems I see most often are unneeded memory copying. If you have enough data to saturate a GPU, then you will also be in a different regime than most things I see (I see strong scaled problems usually).

In your logs, for OpenMP - what system BLAS are you using? If I am not mistaken, dot and update will end up being a Kokkos Kernel (compiled code), and GEMM will end up being a call to your supplied BLAS library.  Judging by your logs, your BLAS library is not threaded. That is, Inner Product and Update are both the same regardless the number of threads.   For Dot, it doesn't matter too much that it is Kokkos vs TPL - I haven't see a big difference between the TPL dots and KK dots... the compiler seems to optimize that type of operation rather well.

**Comment by kyungjoo-kim** [comment]
Hi @ctanis,

My PR is merged and it includes a few things as follows:
- no more HAVE_POINT_IMPORTER thing; internally it always uses point importer for the apply method and now the class does not have its own distObject functions. If you want to communicate in your own way, you need to rely on point map conversion and an importer with the point maps.
- a bit more gpu friendly apply method using atomic which will improve the kernel performance.
- detailed descriptions and performance bench compared with a previous version, see the following read me. 
[read me](https://github.com/trilinos/Trilinos/blob/develop/packages/tpetra/core/example/BlockCrs/README.md)

Hopely this can help your code and understanding each other.

**Comment by ctanis** [comment]
@kyungjoo-kim thanks for your work on this!  I've been traveling, but will see if your changes improve things when I get the chance.  I suspect that the triangular solve that @mhoemmen mentioned is the biggest bottleneck I'm facing.

**Comment by mhoemmen** [comment]
@ctanis I plan to add MKL and cuSPARSE support for block sparse triangular solve -- just need to take care of a few other things first.

**Comment by ctanis** [comment]
@mhoemmen that's great news!

**Comment by ctanis** [comment]
@mhoemmen Can't believe it's been a year since this back-and-forth, but there have been a lot of distractions around here, and my project is ramping back up.  Did you happen to improve the block sparse triangular solve?  I saw another "stale" issue that looked related here on Github.

**Comment by mhoemmen** [comment]
@ctanis I'm not a Trilinos developer any more; in fact, I've changed jobs and now work for [Stellar Science](https://www.stellarscience.com/).  @kyungjoo-kim , @jennloe , and @jjellio are good contacts.  Best wishes with your project!

**Comment by ctanis** [comment]
@mhoemmen congratulations!  i'm slightly familar with stellar science.  do any of the rest of you happen to know whether the block triangle solve mentioned earlier is using kokkos properly yet?  @kyungjoo-kim @jennloe @jjellio

**Comment by csiefer2** [comment]
Well, we're currently hunting down a bug in it :)

**Comment by mhoemmen** [comment]
I forgot to mention @csiefer2 , who is both an expert, and manages to remain my friend despite knowing me for 20 years ; - P

**Comment by ctanis** [comment]
ah cool! @csiefer2 is this a bug I can follow somehow?

**Comment by ctanis** [comment]
I have a new variation on this same problem.  I'm posting it here since it seems to have a lot in common with my earlier questions, but I can create a new issue if you think that's best.  @jjellio this seems particularly up your alley...

I am now using a  straight up CrsMatrix (not a block matrix) and I'm exclusively using Kokkos::OpenMPl on a single MPI process -- no Cuda or anything like that.

Belos does not scale with the number of threads at all.  I am using openblas, which I believe is threaded.
I'm attaching timing output for 4 and 8 processes.  In particular, the orthogonalization and Prec*x don't seem to use threads at all.  (This is RILUK, but Relaxation shows the same behavior). Note that the same trend is visible when preconditioner  is  not used. How do I debug this?  

[belos-timing.txt](https://github.com/trilinos/Trilinos/files/5215651/belos-timing.txt)

Thanks.

**Comment by jennloe** [comment]
When you say 4 and 8 processes, do you mean OpenMP threads?

**Comment by jjellio** [comment]
Hey Craig,

Your logs seem to show the matvec showing some improvement, and the other Linear Algebra/preconditioner stinking.

Threaded BLAS has always been ho-hum to me.  You need a pretty large chunk of data for a ddot/norm to benefit from threads.  I assume you are setting OMP_PLACES=cores or threads (I'd use cores), and OMP_PROC_BIND=spread or close - If you're just playing around with non-MPI launches to test, you can still get some benefits from using `mpirun --map-by ppr:1:socket --bind-to cores` (force the process to bind to a socket rather than bounce around - which will happen if you just `./run_it.exe` ).  We can get more clever on the MPIRUN/OpenMP variables (so they match exactly with the thread counts)


If you are comfortable using KokkosP (profiler) the stacked kernel timer one will print out the actual kernel timings. It also helps to distinguish which kernels are Kokkos vs library blas.

**Comment by ctanis** [comment]
@jennloe yes, i meant openmp threads -- sorry for the miscommunication.

@jjellio I  agree, it looks like Op*x but not Prec*x are improving with threads.  I have experienced with OMP_PLACES and OMP_PROC_BIND settings, but it's as if most of Belos and Ifpack2 are simply not trying to use threads.  I have tried this on fairly large matrices, so it's not a lack of data.

I will look into KokkosP.

BTW, I have built trilinos without MPI support and am using a serial communicator here, since we're trying to drop it into an existing otherwise serial program and don't want to add that dependency.  Maybe I should be using Kokkos kernels?  Do I need to explicitly enable that package?  I see it's replacing blas, at least to some extent.

**Comment by jjellio** [comment]
KokkosKernels should be enabled by default (I don't know if Tpetra will even build if it were disabled).

I highly recommend using KokkosP with the stacked kernel logger.  That will give you a tree-like view of what is getting called (and where).

**Comment by ctanis** [comment]
Just to be really clear about something:  this stuff should be thread-parallel, right?  I mean, Belos+Tpetra+Ifpack2 seems to be  a pretty common combination!

**Comment by jennloe** [comment]
It's probably also worth checking that Kokkos has OpenMP space enabled.  The CMake option is `KOKKOS_ENABLE_OPENMP`.

The Simple Kernel Timer for KokkosP gives less detailed information than the stacked kernel timer, but it may be an easier starting point for timings: https://github.com/kokkos/kokkos-tools/wiki/SimpleKernelTimer

**Comment by ctanis** [comment]
I see already that the "default node" used in Tpetra is OpenMp/HostSpace.

**Comment by ctanis** [comment]
Any idea why exporting KOKKOS_PROFILE_LIBRARY would have no effect?

**Comment by ctanis** [comment]
Using the simple timer, it looks like most of the ICGS and Prec*x calls couldn't possibly be using any kokkos kernels, since the timer says I spent .204 total in all kernels, and all those line items exceed that amount:

[overview.txt](https://github.com/trilinos/Trilinos/files/5220758/overview.txt)

[ip-172-16-1-25-36677-0.txt](https://github.com/trilinos/Trilinos/files/5220760/ip-172-16-1-25-36677-0.txt)

**Comment by jjellio** [comment]
Look for kernel calls without a 'parallel for' or (scan/reduce)

I don't know what Axpby:S12 means, but whatever it is, it is a parallel for.
```
       {
           "kernel-name"    : "KokkosBlas::Axpby::S12",
           "region"         : "",
           "call-count"     : 2,
           "total-time"     : 0.000010,
           "time-per-call"  :       0.00000501,
           "kernel-type"    : "PARALLEL-FOR",
       },
       {
           "kernel-name"    : "KokkosBlas::Axpby::S15",
           "region"         : "",
           "call-count"     : 7,
           "total-time"     : 0.000034,
           "time-per-call"  :       0.00000487,
           "kernel-type"    : "PARALLEL-FOR",
       },
       {
           "kernel-name"    : "KokkosBlas::Mult::S1",
           "region"         : "",
           "call-count"     : 621,
           "total-time"     : 0.005300,
           "time-per-call"  :       0.00000853,
           "kernel-type"    : "PARALLEL-FOR",
       },
```

I am surprised to see Mult being a for loop - it was a GEMM or GEMV call.
I recommend the 'space time' profiler - it will put things into a 'tree' like structure, so you can see which exact regions are slowing down.

Just a glance at the log file, it seems all of your kernel calls are hitting Kokkos (not the TPL blas).

**Comment by brian-kelley** [comment]
@jjellio "S12" is the special case of axpby where alpha is arbitrary and beta is 0.

**Comment by ctanis** [comment]
This is dominating my runtime:  KokkosBlas::gemm[TPL_BLAS,double]  and it does not change at all as I mess with OMP_NUM_THREADS.

It looks like this calls the TPL version (in my case openblas).  How can I get Kokkos to use it's own blas?  Trying to configure trilinos without TPL_BLAS breaks a TeuchosNumerics dependency.

**Comment by nmhamster** [comment]
@ctanis - can you compile OpenBLAS with `USE_OPENMP=1` so you get a threaded BLAS install?

**Comment by ctanis** [comment]
@nmhamster worked like a charm.  it's speeding up pretty good now.. what an ordeal!

Is this really the desirable solution?  Or should I try to get KokkosBlas to work?  I feel like my trilinos configure script must be wacky, since I had to add that flag to enable Kokkos profiling.

anyway, thanks everybody.

**Comment by nmhamster** [comment]
@ctanis - I think either are fine, I personally use the optimized platform BLAS libraries when they are available (e.g. OpenBLAS, MKL, Arm-PL etc) because they should be higher performance than our implementations (although sometimes they aren't). We provide Kokkos versions for compatibility and as a back-stop as not all vendor libraries implement everything we need.

**Comment by jhux2** [comment]
@nmhamster Nice catch.  Is it possible to detect during Trilinos configure that OpenBLAS hasn't been configured properly?

**Comment by nmhamster** [comment]
@jhux2 - if we did this we would do an OpenMP enabled + threaded OpenBLAS check? One option is to check whether `libgomp` or appropriate runtime is needed for linking but that is a little clunky.

**Comment by jhux2** [comment]
Yes, I think that's right (OpenMP + threaded OpenBLAS), so that users would know right away that there's a problem. Checking for a library after the fact isn't great.  What would be nice is some query tool about OpenBLAS's configuration.

**Comment by jjellio** [comment]
I don't think there is a good solution to this - Si is proposing a very specific check, but the problem will manifest on ARM and MKL as well.  I hate to say it, but the burden needs to be on the installer of the blas libs to annotate what they've installed.  

Perhaps a more robust version of a specific check, would be to put the blas library through `nm` or `objdump` and grep for some OpenMP symbol. That would work for both static or shared (it would be weak), and it wouldn't be OpenBLAS specific.  That said, I suspect that kinda check will cause more grief than benefit.

**Comment by jhux2** [comment]
@jjellio You make a reasonable argument, and I agree, there doesn't seem to be a good robust solution on our side.

**Comment by crtrott** [comment]
Regarding the LIBDL thing: Kokkos_ENABLE_PROFILING has been removed, essentially you can't disable it. The LIBDL thing is needed to use KokkosP via dynamically opened libraries - there is also an option to set the callbacks from within the code which works without it. The thing is: I thought libdl should be a default dependency, and if cmake can find it on its own it will enable it unless you explicitly turn it off?

**Comment by jewatkins** [comment]
> Regarding the LIBDL thing: Kokkos_ENABLE_PROFILING has been removed, essentially you can't disable it. The LIBDL thing is needed to use KokkosP via dynamically opened libraries - there is also an option to set the callbacks from within the code which works without it. The thing is: I thought libdl should be a default dependency, and if cmake can find it on its own it will enable it unless you explicitly turn it off?

@crtrott I think this just hit me. I've been nightly auto-profiling and I just realized today that the profiler hasn't been working since 8/25. This seems to coincide with the kokkos 3.2.00 promotion. Will try the LIBDL option.

**Comment by tasmith4** [comment]
@crtrott @jewatkins I hit a similar problem with profilers not working as well.  Some PRs are in progress for both Kokkos and Trilinos:
https://github.com/trilinos/Trilinos/pull/8108
https://github.com/kokkos/kokkos/pull/3424
As per some of the comments on those issues, looks like it might not be quite as simple as setting `-DKokkos_ENABLE_LIBDL=ON` explicitly.

**Comment by mhoemmen** [comment]
begone autobot

**Comment by csiefer2** [comment]
@tasmith4 @ctanis Did this ever get resolved?

**Comment by tasmith4** [comment]
@csiefer2 the fix for the Kokkos profiling TPLs merged in #8233.  @ctanis will have to weigh in on whether his performance issues are now resolved.

**Comment by ctanis** [comment]
@csiefer2 sorry, but i haven't worked on this for years.  it was not fixed when last i looked

**Comment by csiefer2** [comment]
@ctanis Well if this isn't important for you any more, we can close the issue.  We've got a bunch of new work on blockcrs in flight right now and we can generate a new issue for that if people still see performance problems.
